File: amulet/game/_universal/_biome.py
from amulet.biome import Biome
from amulet.game.abc import BiomeData
from amulet.version import VersionNumber


class UniversalBiomeData(BiomeData):
    def to_universal(self, biome: Biome) -> Biome:
        if not self._game_version.supports_version(biome.platform, biome.version):
            raise ValueError("The block is not compatible with this version")
        return biome

    def from_universal(
        self, target_platform: str, target_version: VersionNumber, biome: Biome
    ) -> Biome:
        if not self._game_version.supports_version(biome.platform, biome.version):
            raise ValueError("The block is not compatible with this version")
        return biome

File: src/amulet/_game/abc/biome.cpp
#include "biome.hpp"

namespace Amulet {

BiomeData::BiomeData(py::object biome_data)
    : _biome_data(std::make_unique<py::object>(biome_data))
{
}

BiomeData::~BiomeData()
{
    py::gil_scoped_acquire gil;
    _biome_data = nullptr;
}

Biome BiomeData::translate(const std::string& platform, const VersionNumber& version, const Biome& biome)
{
    py::gil_scoped_acquire gil;
    return _biome_data->attr("translate")(platform, py::cast(version), py::cast(biome)).cast<Biome>();
}

} // namespace Amulet
